src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,161 @@
+            current.previous.next = newNode;
+            current.previous = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element: list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        outOfBoundsCheck(index);
+        Node<T> current = getNodeByIndex(index);
+        return (T) current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        outOfBoundsCheck(index);
+        Node<T> current = getNodeByIndex(index);
+        T removedValue = current.value;
+        current.value = value;
+        return removedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        outOfBoundsCheck(index);
+        if (index == 0) {
+            T removedElement = (T) head.value;
+            head = head.next;
+            size--;
+            return removedElement;
+        } else if (index == size - 1) {
+            T removedElement = (T) tail.value;
+            tail = tail.previous;
+            size--;
+            return removedElement;
+        }
+        T removedElement;
+        Node<T> current = getNodeByIndex(index);
+        removedElement = (T) current.value;
+        unlink(current);
+        size--;
+        return removedElement;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (current.value != null && current.value.equals(object) || current.value == object) {
+                unlink(current);
+                size--;
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void outOfBoundsCheck(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayIndexOutOfBoundsException("Incorrect index: " + index);
+        }
+    }
+
+    private void unlink(Node current) {
+        if (current == head) {
+            head = head.next;
+        } else if (current == tail) {
+            tail = tail.previous;
+        } else {
+            current.previous.next = current.next;
+            current.next.previous = current.previous;
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> current;
+        if (index <= size / 2) {
+            current = head;
+        } else {
+            current = tail;
+        }
+        if (current == head) {
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            for (int i = size - 1; i > index; i--) {
+                current = current.previous;
+            }
+        }
```

Comment:
```suggestion
        if (index < size / 2) {
         current = head;
            for (int i = 0; i < index; i++) {
                current = current.next;
            }
        } else {
            current = tail;
            for (int i = size - 1; i > index; i--) {
                current = current.previous;
            }
        }
```
you can simplify  it